File: 0665-non-decreasing-array/0665-non-decreasing-array.py
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        pre = -inf
        mistake = -1

        for i, n in enumerate(nums):
            if n < pre:
                if mistake >= 0:
                    return False
                mistake = i
                pre = -inf
            else:
                pre = n
        
        if mistake < 0 or mistake == len(nums) - 1: return True

        n0 = nums[mistake - 2] if mistake >= 2 else -inf
        n1 = nums[mistake - 1]
        n2 = nums[mistake]
        n3 = nums[mistake + 1]

        if n0 > n3: return False

        if not (n0 <= n2 <= n3 or n0 <= n1 <= n3): return False

        return True

File: 0665-non-decreasing-array/0665-non-decreasing-array.cpp
class Solution {
public:
    bool checkPossibility(vector<int>& nums) {
        if (nums.size() <= 2) return true;

        bool used = false;

        for (int i = 0; i < nums.size() - 2; i++) {
            bool inv1 = nums[i] <= nums[i + 1];
            bool inv2 = nums[i + 1] <= nums[i + 2];
            bool inv3 = nums[i] <= nums[i + 2];

            if (inv1 && inv2 && inv3) continue;

            if (!inv1 && !inv2 && !inv3) return false;

            if (used) return false;

            if (inv3) {
                nums[i + 1] = nums[i];
                used = true;
            } else if (inv1) {
                nums[i + 2] = nums[i + 1];
                used = true;
            } else {
                used = true;
            }
        }

        return true;
    }
};